File: src/domain/note.ts
/**
 * 笔记领域：实体 + 不变量。
 *
 * `Note` 类型由后端 Rust 结构体经 **specta** 生成（见 `./generated.ts`），
 * 是前后端共享的单一事实来源；本文件只保留校验函数。
 * 修改笔记字段请改后端 `src-tauri/src/domain/note.rs` 后重新生成。
 */

/** 笔记实体（由 specta 从 Rust `Note` 生成）。 */
export type { Note } from "./generated";

/** 内容最大长度（与后端 `CONTENT_MAX_LEN` 一致）。 */
export const CONTENT_MAX_LEN = 5000;

/** 校验笔记内容不变量；返回错误消息，合法则返回 `null`。 */
export function validateNoteContent(content: string): string | null {
  const trimmed = content.trim();
  if (!trimmed) return "笔记内容不能为空。";
  if ([...trimmed].length > CONTENT_MAX_LEN) {
    return `笔记内容不能超过 ${CONTENT_MAX_LEN} 个字符。`;
  }
  return null;
}

File: src-tauri/src/domain/note.rs
//! 笔记实体：内容不变量 + 变更动作。
//!
//! 与任务实体相同的约定：身份与创建时间在构造时生成，字段私有只读，变更必须
//! 经由 [`Note::update`]。

use crate::domain::error::DomainError;
use crate::shared::time;
use serde::Serialize;
use uuid::Uuid;

/// 内容最大长度（字符数，与前端一致）。
pub const CONTENT_MAX_LEN: usize = 5000;

/// 附属于任务（或独立）的笔记聚合。
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Note {
    id: String,
    task_id: Option<String>,
    content: String,
    created_at: String,
}

impl Note {
    /// 构造一个新笔记，就地校验内容不变量。
    ///
    /// 内容会先 `trim`，空白视为无效；`task_id` 可为空（独立笔记）。
    pub fn new(task_id: Option<String>, content: &str) -> Result<Self, DomainError> {
        let content = content.trim();
        validate_content(content)?;
        Ok(Self {
            id: Uuid::new_v4().to_string(),
            task_id,
            content: content.to_owned(),
            created_at: time::utc_now_rfc3339(),
        })
    }

    /// 从持久化原始数据重建实体（仅供仓储适配器使用，跳过校验）。
    pub fn rebuild(id: String, task_id: Option<String>, content: String, created_at: String) -> Self {
        Self {
            id,
            task_id,
            content,
            created_at,
        }
    }

    /// 更新内容：提供的值会重新校验并 `trim`。
    pub fn update(&mut self, content: Option<&str>) -> Result<(), DomainError> {
        if let Some(content) = content {
            let content = content.trim();
            validate_content(content)?;
            self.content = content.to_owned();
        }
        Ok(())
    }

    pub fn id(&self) -> &str {
        &self.id
    }

    pub fn task_id(&self) -> Option<&str> {
        self.task_id.as_deref()
    }

    pub fn content(&self) -> &str {
        &self.content
    }

    pub fn created_at(&self) -> &str {
        &self.created_at
    }
}

fn validate_content(content: &str) -> Result<(), DomainError> {
    if content.is_empty() {
        return Err(DomainError::NoteContentEmpty);
    }
    if content.chars().count() > CONTENT_MAX_LEN {
        return Err(DomainError::NoteContentTooLong);
    }
    Ok(())
}
